### PR TITLE
(wip) feat: add padded locks for striped locking

### PR DESCRIPTION
**Pull request by derrandz**
## Overview

[Context](https://github.com/celestiaorg/celestia-node/pull/2612#discussion_r1322271835)

This PR introduces `PaddedLock` structured to avoid false sharing phenomena for striped locks

## TODO

- [ ] Include benchmarks

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: cmd/celestia/rpc.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ import (
 	"github.com/celestiaorg/celestia-node/state"
 )
 
-const authEnvKey = "CELESTIA_NODE_AUTH_TOKEN" //nolint:gosec
```
Comment (distractedm1nd) [inline cmd/celestia/rpc.go:26]: Pls fix

File: libs/utils/padded_lock.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"sync"
+
+	"golang.org/x/sys/cpu"
+)
+
+// PaddedLock is -as the name indicates- a lock which is padded in memory
+// exactly `cpu.CacheLinePad` length.
+// This is used for stripped locking to avoid-false sharing scenarios.
```
Comment (distractedm1nd) [inline libs/utils/padded_lock.go:11]:
```suggestion
// This is used for striped locking to avoid-false sharing scenarios.
```

**Review by distractedm1nd — CHANGES_REQUESTED** [review]

**Comment by distractedm1nd** [comment]
@derrandz status? looking forward for this

**Comment by derrandz** [comment]
@distractedm1nd should continue on it this weekend, thanks for following up!

**Comment by derrandz** [comment]
@distractedm1nd Can you check up on the benchmark I added (_for starters_) for the eds store... the results I am finding don't show much improvement vs using regular mutexes, I find that weird. Would it be because it's eventually an I/O bound op and not a CPU op?

Will add benchmarks to the rest asap and see the difference, for now, here are the store's:

(_Tried a few homegrown parallelism approaches in benchmarking using wait groups and multiple routines before finally using `RunParallel`, not much difference_)

```
BenchmarkStore

Case 1)

- square size = 128
- parallelism = GOMAXPROCS

Padded stripped lock:
  BenchmarkStore/bench_put_128_parallel-2         	       2	 562190188 ns/op
  BenchmarkStore/bench_put_128_parallel-4         	       3	 472831820 ns/op
  BenchmarkStore/bench_put_128_parallel-8         	       2	 514809125 ns/op

Mutex stripped lock:

    BenchmarkStore/bench_put_128_parallel-2         	       2	 543366292 ns/op
    BenchmarkStore/bench_put_128_parallel-4         	       3	 470625778 ns/op
    BenchmarkStore/bench_put_128_parallel-8         	       3	 467839722 ns/op


- square size = 128
- parallelism = 200

PaddedLocks:

BenchmarkStore/bench_put_128_parallel-2         	       2	 541508938 ns/op
BenchmarkStore/bench_put_128_parallel-4         	       3	 485055959 ns/op
BenchmarkStore/bench_put_128_parallel-8         	       3	 468695945 ns/op

Mutex:

BenchmarkStore/bench_put_128_parallel-2         	       2	 535595584 ns/op
BenchmarkStore/bench_put_128_parallel-4         	       3	 462523208 ns/op
BenchmarkStore/bench_put_128_parallel-8         	       3	 475363972 ns/op


- square size = 128
- parallelism = 512

PaddedLocks:
BenchmarkStore/bench_put_128_parallel-2         	       2	 520852646 ns/op
BenchmarkStore/bench_put_128_parallel-4         	       3	 468282972 ns/op
BenchmarkStore/bench_put_128_parallel-8         	       2	 517932875 ns/op

Mutex:

BenchmarkStore/bench_put_128_parallel-2         	       2	 520831520 ns/op
BenchmarkStore/bench_put_128_parallel-4         	       3	 477697167 ns/op
BenchmarkStore/bench_put_128_parallel-8         	       3	 463459944 ns/op
```

**Comment by derrandz** [comment]
Update:

Here are benchmarks on an i9 machine. Although the theory behind this optimization makes sense, in practice, there isn't much performance gains. So I think I'll close this PR for now.

```
                                  │ benchmarks.mutex.txt │       benchmarks.padded.txt        │
                                  │        sec/op        │    sec/op     vs base              │
Store/bench_put_128-2                       1.856 ±  30%    1.898 ± 23%       ~ (p=0.818 n=6)
Store/bench_put_128-4                       1.788 ±  12%    1.761 ±  9%       ~ (p=0.818 n=6)
Store/bench_put_128-6                       2.021 ±  16%    1.834 ± 14%       ~ (p=0.310 n=6)
Store/bench_put_128-8                       2.148 ±  19%    2.036 ± 12%       ~ (p=0.485 n=6)
Store/bench_put_128-10                      2.194 ±  40%    2.179 ±  9%       ~ (p=0.699 n=6)
Store/bench_read_128-2                     290.8m ± 273%   334.4m ± 20%       ~ (p=0.818 n=6)
Store/bench_read_128-4                     197.9m ±  17%   187.4m ± 12%       ~ (p=0.240 n=6)
Store/bench_read_128-6                     147.9m ±   8%   151.9m ± 15%       ~ (p=0.699 n=6)
Store/bench_read_128-8                     120.4m ±  29%   123.2m ± 14%       ~ (p=0.589 n=6)
Store/bench_read_128-10                    127.3m ±   7%   130.5m ± 35%       ~ (p=0.240 n=6)
Store/bench_put_128_concurrent-2            2.103 ±  43%    2.511 ± 23%       ~ (p=0.180 n=6)
Store/bench_put_128_concurrent-4            1.334 ±  24%    1.345 ± 44%       ~ (p=0.818 n=6)
Store/bench_put_128_concurrent-6            1.113 ±  28%    1.115 ± 18%       ~ (p=0.818 n=6)
Store/bench_put_128_concurrent-8          1005.6m ±  23%   908.0m ± 20%       ~ (p=0.589 n=6)
Store/bench_put_128_concurrent-10          944.6m ±  13%   935.9m ±  4%       ~ (p=1.000 n=6)
geomean                                    745.6m          749.8m        +0.57%
```

**Comment by derrandz** [comment]
I hate to see this go, but it is what it is. Closing for now

cc @distractedm1nd @Wondertan

**Comment by derrandz** [comment]
cc @distractedm1nd @Wondertan 

Will revaluate the actual need for this PR by providing an analysis using: https://grafana.com/docs/pyroscope/latest/configure-client/language-sdks/go_push/#mutex-profiling

This will be helpful in identifying all sources of contention within the app. This is going to be more insightful than just the theoretical reasons I've provided.

I will also revive this PR after getting some more accurate benchmarks on a server with a few hundered thousand runs... If pyroscope proves that there is a contention point indeed.

**Pull request closed without merging**